server/protocols/jmap/src/main/java/org/apache/james/jmap/model/CreationMessage.java

```diff
@@ -236,12 +232,13 @@ public CreationMessage build() {
                     .fromMap(map));
         }
 
-        private Optional<OldKeyword> getOldKeywords() {
-            if (isAnswered.isPresent() || isFlagged.isPresent() || isUnread.isPresent() || isDraft.isPresent() || isForwarded.isPresent()) {
-                return Optional.of(new OldKeyword(isUnread, isFlagged, isAnswered, isDraft, isForwarded));
-            }
-            return Optional.empty();
+        public Keywords computeKeywords(Optional<Keywords> keywords, Optional<OldKeyword> oldKeyword) {
+            return oldKeyword.map(OldKeyword::asKeywords)
+                .map(Optional::of)
```

Comment:
if you use Optional::of it's not null, so it's not empty, so next line is useless, no?

Reply:
Not at all!

Say it in english:

 - We want to use the OldKeywords if they exist
 - Or else the keywords, if they exist
 - Or else the empty keywords ....

When I say it in english does something looks not needed? Redoundant? That double optional wrapping could be replaced by a if if you concider it tooooo strange.

Reply:
Note:

Reversing a bit the chain makes it more understandable:

```
            return keywords.map(Optional::of)
                .orElse(oldKeyword.map(OldKeyword::asKeywords))
                .orElse(Keywords.DEFAULT_VALUE);
```

(We ensured previously keywords and oldKeywords are not both present...)

Reply:
OptionalUtils.or() would do it...

Reply:
You did not answer my question.
I understand what you want to do, but: Optional.of is never empty, so how could it work?

Reply:
No no no it's very logical and works.

Code snipset:

```
1    keywords
2                .map(Optional::of)
3               .orElse(oldKeyword.map(OldKeyword::asKeywords))
4              .orElse(Keywords.DEFAULT_VALUE);
```

Case 1: keyword is here, no OldKeywords:

```
1  Optional(k]
 2 Optional[Optional[k]]
 3 Optional[k]
 4 k
```

Case 2: keyword is empty, no OldKeywords:

```
1  Optional(]
 2 Optional[]
 3 Optional[] // OldKeywords is also empty
 4 Default
```

Case 3: keyword is here,  OldKeywords is here:

```
1  Optional(]
 2 Optional[]
 3 Optional[o]
 4 o
```